### PR TITLE
Remove notion of Modification

### PR DESCRIPTION
**Pull request by ifesdjeen**
Patch for CASSANDRA-19117 by Alex Petrov; reviewed by Abe Ratnofsky

### CODE REVIEW

File: harry-core/src/harry/core/Configuration.java [hunk 1]
```diff
@@ -885,20 +885,20 @@ public DefaultCDSelectorConfiguration build()
...
-        public final DistributionConfig rows_per_modification;
+        public final DistributionConfig operations_per_lts;
         public final int max_partition_size;
         public final Map<OpSelectors.OperationKind, Integer> operation_kind_weights;
         public final Map<OpSelectors.OperationKind, long[]> column_mask_bitsets;
 
         @JsonCreator
         public DefaultCDSelectorConfiguration(@JsonProperty("modifications_per_lts") DistributionConfig modifications_per_lts,
-                                              @JsonProperty("rows_per_modification") DistributionConfig rows_per_modification,
+                                              @JsonProperty("rows_per_modification") DistributionConfig operations_per_lts,
```
Comment (maedhroz) [inline harry-core/src/harry/core/Configuration.java]: Leaving "rows_per_modification" for back-compat?

File: harry-core/src/harry/core/Configuration.java [hunk 2]
```diff
@@ -885,20 +885,20 @@ public DefaultCDSelectorConfiguration build()
     public static class DefaultCDSelectorConfiguration implements CDSelectorConfiguration
     {
         public final DistributionConfig modifications_per_lts;
```
Comment (maedhroz) [inline harry-core/src/harry/core/Configuration.java]: Is `modifications_per_lts` here unused now?

File: harry-core/src/harry/dsl/HistoryBuilder.java [hunk 3]
```diff
@@ -361,32 +350,22 @@ public HistoryBuilder finish()
 
         void addSteps(List<Step> steps)
         {
-            List<ReplayingVisitor.Batch> batches = new ArrayList<>();
-            Counter m = new Counter();
+            List<ReplayingVisitor.Operation> operations = new ArrayList<>();
             Counter opId = new Counter();
             for (Step step : steps)
             {
-                batches.add(step.toBatch(pd, lts, m.get(), opId::getAndIncrement));
```
Comment (maedhroz) [inline harry-core/src/harry/dsl/HistoryBuilder.java:369]: Cool, so we no longer need the notion of a modification ID?

File: harry-core/src/harry/model/sut/injvm/InJVMTokenAwareVisitExecutor.java [hunk 4]
```diff
@@ -72,44 +78,43 @@ public InJVMTokenAwareVisitExecutor(Run run,
...
+                }
+            }
+            catch (Throwable t)
+            {
+                int delaySecs = 1;
+                logger.error(String.format("Caught message while trying to execute: %s. Scheduled to retry in %s seconds", statement, delaySecs), t);
+                Uninterruptibles.sleepUninterruptibly(delaySecs, TimeUnit.SECONDS);
+            }
         }
+        throw new IllegalStateException(String.format("Can not execute statement %s after %d retries", statement, retries));
```
Comment (maedhroz) [inline harry-core/src/harry/model/sut/injvm/InJVMTokenAwareVisitExecutor.java:117]: So tl;dr we made this synchronous and added some logging for visibility? I'll figure out why I suppose as I look at its callers...

File: harry-core/src/harry/model/DescriptorSelectorBuilder.java [hunk 5]
```diff
@@ -33,7 +33,7 @@ public class DescriptorSelectorBuilder implements Configuration.CDSelectorConfig
 {
     private Function<SchemaSpec, OpSelectors.ColumnSelector> columnSelectorFactory;
     private OpSelectors.OperationSelector operationSelector = DEFAULT_OP_SELECTOR;
-    private Distribution numberOfRowsDistribution = new Distribution.ScaledDistribution(2, 30);
+    private Distribution modificationsPerLts = new Distribution.ScaledDistribution(2, 30);
     private Distribution numberOfModificationsDistribution = new Distribution.ScaledDistribution(1, 3);
```
Comment (maedhroz) [inline harry-core/src/harry/model/DescriptorSelectorBuilder.java]: `numberOfModificationsDistribution` is unused now?

File: harry-core/src/harry/model/DescriptorSelectorBuilder.java [hunk 6]
```diff
@@ -72,9 +72,9 @@ public DescriptorSelectorBuilder setOperationSelector(OpSelectors.OperationSelec
      * In a given modification, we are only able to visit as many rows as there are rows in the partition, so
      * we'll always be limited by this.
      **/
-    public DescriptorSelectorBuilder setRowsPerModificationDistribution(Distribution numberOfRowsDistribution)
+    public DescriptorSelectorBuilder setModificationsPerLts(Distribution numberOfRowsDistribution)
```
Comment (maedhroz) [inline harry-core/src/harry/model/DescriptorSelectorBuilder.java]:
```suggestion
    public DescriptorSelectorBuilder setModificationsPerLts(Distribution this.modificationsPerLts)
```
?

File: harry-core/src/harry/model/DescriptorSelectorBuilder.java [hunk 7]
```diff
@@ -72,9 +72,9 @@ public DescriptorSelectorBuilder setOperationSelector(OpSelectors.OperationSelec
      * In a given modification, we are only able to visit as many rows as there are rows in the partition, so
      * we'll always be limited by this.
```
Comment (maedhroz) [inline harry-core/src/harry/model/DescriptorSelectorBuilder.java]: nit: Comment needs tweaking?

File: harry-core/src/harry/model/DescriptorSelectorBuilder.java [hunk 8]
```diff
@@ -33,7 +33,7 @@ public class DescriptorSelectorBuilder implements Configuration.CDSelectorConfig
 {
     private Function<SchemaSpec, OpSelectors.ColumnSelector> columnSelectorFactory;
     private OpSelectors.OperationSelector operationSelector = DEFAULT_OP_SELECTOR;
-    private Distribution numberOfRowsDistribution = new Distribution.ScaledDistribution(2, 30);
+    private Distribution modificationsPerLts = new Distribution.ScaledDistribution(2, 30);
     private Distribution numberOfModificationsDistribution = new Distribution.ScaledDistribution(1, 3);
     private int maxPartitionSize = Integer.MAX_VALUE;
     private Function<SchemaSpec, int[]> fractionsSupplier = null;
```
Comment (maedhroz) [inline harry-core/src/harry/model/DescriptorSelectorBuilder.java]: nit: What does `fractionsSupplier` do?

File: harry-core/src/harry/model/DescriptorSelectorBuilder.java (hunk 6 above)
Comment (maedhroz) [reply harry-core/src/harry/model/DescriptorSelectorBuilder.java]: I guess we've named it `operationsPerLtsDistribution` elsewhere, so I guess that's fine too if we want to be explicit.

File: harry-core/test/harry/model/OpSelectorsTest.java [hunk 9]
```diff
@@ -73,19 +72,16 @@ public void testRowDataDescriptorSupplier()
         for (int lts = 0; lts < RUNS; lts++)
         {
             long pd = pdSupplier.pd(lts);
-            for (int m = 0; m < descriptorSelector.numberOfModifications(lts); m++)
+
+            int opsPerMod = descriptorSelector.operationsPerLts(lts);
```
Comment (maedhroz) [inline harry-core/test/harry/model/OpSelectorsTest.java]:
```suggestion
            int opsPerLts = descriptorSelector.operationsPerLts(lts);
```

File: harry-core/test/harry/operations/RelationTest.java [hunk 10]
```diff
@@ -183,12 +183,12 @@ public BitSet columnMask(long pd, long lts, long opId, OpSelectors.OperationKind
...
                                                                       }
 
-                                                                      public long rowId(long pd, long lts, long cd)
+                                                                      public long opId(long pd, long lts, long cd)
                                                                       {
                                                                           return 0;
                                                                       }
 
-                                                                      public long modificationId(long pd, long cd, long lts, long vd, int col)
+                                                                      public long operationId(long pd, long cd, long lts, long vd, int col)
```
Comment (maedhroz) [inline harry-core/test/harry/operations/RelationTest.java]: unused now?

**Review by maedhroz — COMMENTED** [review]
Mostly looks good. Just dropped some minor comments/cleanup requests.

Is `CDSelectorConfigurationBuilder` going to retain `modifications_per_lts` and `rows_per_modification`?

File: harry-core/src/harry/core/Configuration.java (hunk 1 above)
Comment (ifesdjeen) [reply harry-core/src/harry/core/Configuration.java]: Removed rows_per_modification now and cleaned up modifications_per_lts. Missed this.

File: harry-core/src/harry/core/Configuration.java (hunk 2 above)
Comment (ifesdjeen) [reply harry-core/src/harry/core/Configuration.java]: Yes; removed

File: harry-core/src/harry/dsl/HistoryBuilder.java (hunk 3 above)
Comment (ifesdjeen) [reply harry-core/src/harry/dsl/HistoryBuilder.java:369]: yes! Is a small simplificaiton, but makes things much simpler when working on history builder, less indirection.

File: harry-core/src/harry/model/sut/injvm/InJVMTokenAwareVisitExecutor.java (hunk 4 above)
Comment (ifesdjeen) [reply harry-core/src/harry/model/sut/injvm/InJVMTokenAwareVisitExecutor.java:117]: Let's say this was the simplest way to know that a query _has_ in fact finished. This is a barebone model, and probably should be improved upon. But at least it gives you some ground to stand upon.

File: harry-core/src/harry/model/DescriptorSelectorBuilder.java (hunk 5 above)
Comment (ifesdjeen) [reply harry-core/src/harry/model/DescriptorSelectorBuilder.java]: Good point. Removed!

File: harry-core/src/harry/model/DescriptorSelectorBuilder.java (hunk 6 above)
Comment (ifesdjeen) [reply harry-core/src/harry/model/DescriptorSelectorBuilder.java]: Renamed to `setOperationsPerLts`

File: harry-core/src/harry/model/DescriptorSelectorBuilder.java (hunk 8 above)
Comment (ifesdjeen) [reply harry-core/src/harry/model/DescriptorSelectorBuilder.java]: The initial idea was to use HierarchicalDescriptorSelector: sort of to make "predictable" partitions, where parts of CK would repeat.

File: harry-core/test/harry/model/OpSelectorsTest.java (hunk 9 above)
Comment (ifesdjeen) [reply harry-core/test/harry/model/OpSelectorsTest.java]: Great point!

File: harry-core/test/harry/operations/RelationTest.java (hunk 10 above)
Comment (ifesdjeen) [reply harry-core/test/harry/operations/RelationTest.java]: Good point! Removed this and some more.

**Comment by ifesdjeen** [comment]
Removed modifications_per_lts and rows_per_modification and introduced operations_per_lts.

I _think_ most of your comments should now be addressed.

**Review by maedhroz — APPROVED** [review]

File: harry-core/src/harry/concurrent/Uninterruptibles.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
+package harry.concurrent;
+
+import java.util.concurrent.TimeUnit;
+
+public class Uninterruptibles
```
Comment (aratno) [inline harry-core/src/harry/concurrent/Uninterruptibles.java:5]: Avoiding using Guava?

**Review by aratno — APPROVED** [review]

**Pull request merged**